STI - clean test pass

GregSutcliffe opened the pull request:
This probably still needs work... but it's a start...

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -137,7 +140,7 @@ def hostgroup_or_environment_selected
       @host        = Host.find(params[:host_id])               if params[:host_id].to_i > 0
       if @environment or @hostgroup
-        @host ||= Host.new
+        @host ||= Host.new :type => params['host']['type']
```
why do you need it? afair, new still accepts type?

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -127,6 +129,7 @@ def compute_resource_selected
   end
 
   def hostgroup_or_environment_selected
+    logger.info params.inspect
```
debug?

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -494,6 +497,11 @@ def find_by_name
   end
 
+  def set_base_type
+    params['host'] ||= {}
```
I'm really not sure why you need this, is it because of tests?

ohadlevy commented on app/helpers/hosts_helper.rb:
```diff
@@ -240,7 +240,8 @@ def host_title_actions(host, vm)
                                   :title => "Trigger a puppetrun on a node; requires that puppet run is enabled"),
             link_to_if_authorized("All Puppet Classes", hash_for_storeconfig_klasses_host_path(:id => host).merge(:auth_action => :read),
-                                  :disabled => host.resources.count == 0,
+                                  :disabled => (host.resources.count == 0 rescue false),
```
if you broke of puppet, then there is no longer resources and we should not show this btn

ohadlevy commented on app/models/host.rb:
```diff
@@ -1,951 +1,11 @@
-require 'facts_importer'
+class Host < ActiveRecord::Base
```
this is confusing..

shouldnt this be host::base ?

GregSutcliffe committed:
Add STI to hosts table
Move core host code to Host::* namespace
Implement Stub "Host" module to avoid changing large amounts of code

This commit also breaks the link between Puppet::Rails::Host and our
Host object. As such the storeconfig support is gone, and the "All
Puppet Classes" button has been removed

GregSutcliffe commented:
Clean test pass, no controller changes, index added to host.type. I think it's good to go! @ohadlevy over to you!

ohadlevy commented on app/helpers/hosts_helper.rb:
```diff
@@ -238,10 +238,7 @@ def host_title_actions(host, vm)
-            link_to_if_authorized("All Puppet Classes", hash_for_storeconfig_klasses_host_path(:id => host).merge(:auth_action => :read),
-                                  :disabled => host.resources.count == 0,
-                                  :title => "Show all host puppet classes, requires storeconfigs")
+                                  :title => "Trigger a puppetrun on a node; requires that puppet run is enabled")
```
it should be clearly stated that this commit also removes the store config integration between foreman & puppet?

GregSutcliffe replied:
Yes, I'll add that comment.

ohadlevy commented on app/models/domain.rb:
```diff
@@ -4,7 +4,7 @@ class Domain < ActiveRecord::Base
   include Taxonomix
 
-  has_many :hosts
+  has_many :hosts, :class_name => 'Host::Managed'
```
I wonder if it makes any sense to make it a mixin - so later on when we change it we don't need to do it again

GregSutcliffe replied:
Sounds sensible. How difficult is that? I want to get this done today so I can enjoy my PTO, if possible :P

ohadlevy commented on app/models/host.rb:
```diff
@@ -1,951 +1,27 @@
+module Host
+
+  def self.method_missing(method, *args, &block)
+    super
```
does it make sense to call super here at all? we already matched a method_missing ? (e.g. if we got here, we already didn't find a method)?

GregSutcliffe replied:
I guess not. Will fix.

ohadlevy commented on app/models/puppetclass.rb:
```diff
@@ -6,7 +6,7 @@ class Puppetclass < ActiveRecord::Base
   has_many :hostgroups, :through => :hostgroup_classes
   has_many :host_classes, :dependent => :destroy
-  has_many :hosts, :through => :host_classes
+  has_many :hosts, :through => :host_classes, :foreign_key => :host_id
```
do we still need the foreign_key ?

GregSutcliffe replied:
Seems fine without it:

```
irb(main):009:0> Host.first.puppetclasses.map{|p| p.try(:name)}
=> ["foreman", "puppet::server"]
```

Clean test run as well. Removing.

ohadlevy commented on app/models/token.rb:
```diff
@@ -12,7 +12,7 @@
 class Token < ActiveRecord::Base
   attr_accessible :value, :expires
-  belongs_to :host
+  belongs_to :host, :foreign_key => :host_id
```
stange, don't we need class_name? otherwise, how does it know the relationship?

GregSutcliffe replied:
Not sure, seems to work though:

```
irb(main):006:0> Host.last.token.to_s
=> "83c29bb2-33cf-401b-bd3c-c03c00562381"
```

ohadlevy commented on app/models/user_fact.rb:
```diff
@@ -2,7 +2,7 @@
 class UserFact < ActiveRecord::Base
   belongs_to :user
-  belongs_to :fact_name, :class_name => "Puppet::Rails::FactName"
+  belongs_to :fact_name, :class_name => "FactName"
```
I think you can drop the class_name here

GregSutcliffe replied:
Clean tests pass without that. Removing from the PR.

GregSutcliffe commented:
@ohadlevy Mixin added (mostly copied from Taxonomix.rb). Seems to work for me, all tests pass and I can create/install libvirt VMs...

GregSutcliffe commented:
Fact upload is broken, investigating....

GregSutcliffe committed:
Include merge_facts from Puppet::Rails:Host since we don;t inherit it now

GregSutcliffe commented:
Fact upload fixed, puppet reports seem to work fine too.

GregSutcliffe committed:
Try to cut down on pointless DB calls to FactName

GregSutcliffe commented:
@ohadlevy latest commit is my attempt at this - I'm sure it can be cleaner. Seems to work though, I get one call to the DB for all the names, followed by the updates to FactValues.

isratrade commented:
@ohadlevy, @GregSutcliffe, I pushed pr #430 which made come changes to Greg's branch.
